### PR TITLE
Malicious PSI

### PR DESCRIPTION
**Pull request by kilyig**
The malicious PSI protocol.

### CODE REVIEW

File: src/psi.rs [hunk 1]
```diff
@@ -275,3 +281,45 @@ fn hash_2(input1: u64, input2: Fq) -> Fq {
...
     hash
 }
+
+/// given x coords construct Li polynomials
+fn construct_lagrange_basis<F: FftField>(evaluation_domain: &[F]) -> Vec<DensePolynomial<F>> {
+    let mut bases = Vec::with_capacity(evaluation_domain.len());
+    for i in 0..evaluation_domain.len() {
+        let mut l_i = DensePolynomial::from_coefficients_slice(&[F::one()]);
+        let x_i = evaluation_domain[i];
+        for (j, _) in evaluation_domain.iter().enumerate() {
```
Comment (rsproule) [inline src/psi.rs]: the mix of 0..len and .enumerate while ignore the value is weird. just be consistent. also in the outer loop, you ARE using the value (`x_i`) so use `.enumerate()`

File: src/psi.rs [hunk 2]
```diff
@@ -275,3 +281,45 @@ fn hash_2(input1: u64, input2: Fq) -> Fq {
...
+    let roots: Vec<_> = (0..n).map(|_| Fr::rand(&mut rng)).collect();
+    
+    let lagrange_basis = construct_lagrange_basis(&roots);
+    let f_evals: Vec<_> = (0..n).map(|_| Fr::rand(&mut rng)).collect();
+
+    let mut f_slow = DensePolynomial::default();
+    for (li, &fi) in lagrange_basis.iter().zip(f_evals.iter()) {
+        f_slow += (fi, li);
+    }
+
```
Comment (rsproule) [inline src/psi.rs]: should have some assertion here? i see u were probably just checking by eye but might as well add assertions so if you break this, the test will fail

File: src/mal_psi.rs [hunk 3]
```diff
@@ -197,62 +191,123 @@ pub fn receiver_2(capital_k: Vec<Fq>, m: Fq, b_i_array: Vec<BigInt<4>>, set_y: &
...
-    // In AES, encryption and decryption are done by the same operation
-    // from: https://docs.rs/aes-gcm-siv/latest/aes_gcm_siv/#usage
-    // also: https://stackoverflow.com/questions/23850486/how-do-i-convert-a-string-into-a-vector-of-bytes-in-rust
-    // TODO: why does it give an error when I run `.decrypt`?
-    // TODO: is is_ok() okay (lol) or should I propagate the error with Result<T, E>?
-    // TODO: test whether applying cipher.encrypt gives you the initial value (how is this possible?)
-    let permuted_bytes = cipher.encrypt(nonce, input_string.as_bytes().as_ref());
-    assert!(permuted_bytes.is_ok());
-    let permuted: Fq = <Fq as PrimeField>::from_le_bytes_mod_order(&permuted_bytes.unwrap());
+fn pi_inverse(elt: Fq) -> Fq {
```
Comment (rsproule) [inline src/mal_psi.rs:225]: this looks the same to me as `pi()`

File: src/mal_psi.rs [hunk 4]
```diff
@@ -197,62 +191,123 @@ pub fn receiver_2(capital_k: Vec<Fq>, m: Fq, b_i_array: Vec<BigInt<4>>, set_y: &
...
 fn hash_1(input: u64) -> Fq {
     // https://docs.rs/sha2/latest/sha2/
     let mut hasher = Sha256::new();
     hasher.update(input.to_le_bytes());
     let result = hasher.finalize();
 
-    // need to shorten it to u64 for now.
-    // represent those 8 bytes as a single u64
+    // map the output of the hash back to a field element
     let hash: Fq = <Fq as PrimeField>::from_le_bytes_mod_order(&result);
```
Comment (rsproule) [inline src/mal_psi.rs:304]: unnecessary lh value

File: src/mal_psi.rs [hunk 5]
```diff
@@ -197,62 +191,123 @@ pub fn receiver_2(capital_k: Vec<Fq>, m: Fq, b_i_array: Vec<BigInt<4>>, set_y: &
...
+///  
+/// # Arguments
+///
+/// * `block` - An array with 16 bytes
+/// 
+/// # Outputs
+///
+/// * `elt` - A field element
+fn block_to_field(block: GenericArray<u8, <Aes128 as BlockSizeUser>::BlockSize>) -> Fq {
+    let elt: Fq = <Fq as PrimeField>::from_le_bytes_mod_order(block.as_slice());
```
Comment (rsproule) [inline src/mal_psi.rs:282]: unnecessary lh assignment

File: tests/malicious_test.rs [hunk 6]
```diff
@@ -11,20 +8,41 @@ fn test_two_small_sets() {
...
     let set_y: Vec<u64> = [3u64, 4u64, 5u64].to_vec();
 
+    run_protocol_and_check(&set_x, &set_y);
+}
+
+fn run_protocol_and_check(set_x: &Vec<u64>, set_y: &Vec<u64>) {
+    let plaintext_intersection: HashSet<u64> = plaintext_intersection(&set_x, &set_y).into_iter().collect();
+    let protocol_intersection: HashSet<u64> = protocol_intersection(&set_x, &set_y).into_iter().collect();
+
+    assert!(plaintext_intersection == protocol_intersection);
```
Comment (rsproule) [inline tests/malicious_test.rs:18]: `assert_eq!`

File: src/mal_psi.rs [hunk 7]
```diff
@@ -197,62 +191,123 @@ pub fn receiver_2(capital_k: Vec<Fq>, m: Fq, b_i_array: Vec<BigInt<4>>, set_y: &
...
+/// * `elt` - A field element
+/// 
+/// # Outputs
+///
+/// * `block` - An array with 16 bytes
+fn field_to_block(elt: Fq) -> GenericArray<u8, <Aes128 as BlockSizeUser>::BlockSize> {
+    let elt_bytes = elt.into_bigint().to_bytes_le();
+
+    // TODO: this is ugly
+    let block = GenericArray::from([
```
Comment (rsproule) [inline src/mal_psi.rs:250]: can construct `fromSlice()`, no?

Comment (kilyig) [reply src/mal_psi.rs:250]: Gives me weird errors but I'll try again

File: src/psi.rs (hunk 1 above)
Comment (kilyig) [reply src/psi.rs]: I agree. I copied the code from `fast-eval` but can still change it.

File: src/mal_psi.rs (hunk 3 above)
Comment (kilyig) [reply src/mal_psi.rs:225]: Same as before: there might be an issue with the merge request.

**Comment by kilyig** [comment]
Malicious PSI done except rng.

**Pull request merged**